cdap-ui/app/cdap/components/DataPrep/store/DataPrepActionCreator.js

```diff
@@ -76,6 +77,81 @@ export function execute(addDirective, shouldReset, hideLoading = false) {
   });
 }
 
+function setWorkspaceRetry(params, observer, workspaceId) {
+  MyDataPrepApi.getWorkspace(params)
+    .subscribe((res) => {
+      let {dataprep} = DataPrepStore.getState();
+      /*
+        1. Open a tab with huge data (like 400 columns and 100 rows)
+        2. Change of mind, open another tab
+        3. 2nd tab's data comes in quick and you are happy browsing it
+        4. Baam 1st tab's data comes and royally overwrites the one you are seeing.
+
+        This is to prevent that. We can't cancel the request we made but we should show only that is relevant
+      */
+      if (dataprep.workspaceId !== workspaceId) {
+        return;
+      }
+      let directives = objectQuery(res, 'values', '0', 'recipe', 'directives') || [];
+      let requestBody = directiveRequestBodyCreator(directives);
+
+      let workspaceUri = objectQuery(res, 'values', '0', 'properties', 'path');
+      let workspaceInfo = objectQuery(res, 'values', '0');
+
+      MyDataPrepApi.execute(params, requestBody)
+        .subscribe((response) => {
+          observer.onNext(response);
+
+          DataPrepStore.dispatch({
+            type: DataPrepActions.setWorkspace,
+            payload: {
+              data: response.values,
+              headers: response.header,
+              types: response.types,
+              directives,
+              workspaceId,
+              workspaceUri,
+              workspaceInfo
+            }
+          });
+
+          fetchColumnsInformation(params, requestBody, response.header);
+
+        }, (err) => {
+          // Backend returned an exception. Show default error message for now able to show data.
+          if (workspaceRetries < 3) {
+            workspaceRetries += 1;
+            setWorkspaceRetry(params, observer, workspaceId);
+          } else {
+            observer.onNext(err);
+            DataPrepStore.dispatch({
+              type: DataPrepActions.disableLoading
+            });
+            DataPrepStore.dispatch({
+              type: DataPrepActions.setDataError,
+              payload: {
+                errorMessage: objectQuery(err, 'message') || objectQuery(err, 'response', 'message') || objectQuery(err, 'response')
+              }
+            });
+          }
+        });
+
+    }, (err) => {
+      if (workspaceRetries < 3) {
+        workspaceRetries += 1;
+        setWorkspaceRetry(params, observer, workspaceId);
+      } else {
+        DataPrepStore.dispatch({
+          type: DataPrepActions.setDataError,
+          payload: {
+            errorMessage: objectQuery(err, 'message') || objectQuery(err, 'response', 'message') || objectQuery(err, 'response')
```

Comment:
If this is not going to be used then do we need this? just set `errorMessage` to `true`? Since we are going to show something from i18n